File: backend-go/api/inquiryEndpoint.go
package api

import (
	"encoding/json"
	"log"
	"superQiMiniAppBackend/alipay"

	"github.com/gofiber/fiber/v2"
)

type applyTokenRequest struct {
	AuthCode string `json:"auth_code" validate:"required"`
}

type inquiryCardsRequest struct {
	AccessToken string `json:"accessToken" validate:"required"`
}

func InitInquiryEndpoint(group fiber.Router) {
	// Endpoint to exchange auth code for access token (specifically for card inquiry)
	group.Post("/users/inquiry-cards/apply-token", func(ctx *fiber.Ctx) error {
		var request applyTokenRequest
		if err := ctx.BodyParser(&request); err != nil {
			return fiber.NewError(fiber.StatusBadRequest, "Invalid request body")
		}

		log.Println("=================================================================")
		log.Println("STARTING TOKEN EXCHANGE FOR CARD INQUIRY")
		log.Println("=================================================================")
		log.Println("[INFO] Auth code received from frontend")

		tokenResponse, err := alipay.Interface.ApplyToken(request.AuthCode)
		if err != nil {
			log.Printf("[ERROR] Token exchange failed: %v\n", err)
			log.Println("=================================================================\n")
			return fiber.NewError(fiber.StatusBadRequest, err.Error())
		}

		tokenResponseJson, _ := json.MarshalIndent(tokenResponse, "", "  ")
		log.Printf("[SUCCESS] Token response received:\n%s\n\n", string(tokenResponseJson))

		if tokenResponse.Result.ResultCode != "SUCCESS" {
			log.Printf("[ERROR] Invalid token response: %s\n", tokenResponse.Result.ResultMessage)
			log.Println("=================================================================\n")
			return fiber.NewError(fiber.StatusBadRequest, "Invalid token response: "+tokenResponse.Result.ResultMessage)
		}

		log.Println("[INFO] Token exchange successful")
		log.Printf("[INFO] Customer ID: %s\n", tokenResponse.CustomerID)
		log.Printf("[INFO] Access token obtained (valid until: %s)\n", tokenResponse.AccessTokenExpiryTime)
		log.Println("[SUCCESS] Returning access token to frontend")
		log.Println("=================================================================\n")

		response := fiber.Map{
			"accessToken":            tokenResponse.AccessToken,
			"customerId":             tokenResponse.CustomerID,
			"accessTokenExpiryTime":  tokenResponse.AccessTokenExpiryTime,
		}

		return ctx.JSON(response)
	})

	// Endpoint to get user card list using access token
	group.Post("/users/inquiry-cards", func(ctx *fiber.Ctx) error {
		var request inquiryCardsRequest
		if err := ctx.BodyParser(&request); err != nil {
			return fiber.NewError(fiber.StatusBadRequest, "Invalid request body")
		}

		log.Println("=================================================================")
		log.Println("STARTING USER CARD LIST INQUIRY")
		log.Println("=================================================================")
		log.Println("[INFO] Access token received from frontend")
		log.Println("[INFO] Calling Alipay+ inquiryUserCardList API...")

		cardListResponse, err := alipay.Interface.InquiryUserCardList(request.AccessToken)
		if err != nil {
			log.Printf("[ERROR] Card inquiry failed: %v\n", err)
			log.Println("=================================================================\n")
			return fiber.NewError(fiber.StatusInternalServerError, err.Error())
		}

		cardListResponseJson, _ := json.MarshalIndent(cardListResponse, "", "  ")
		log.Printf("[SUCCESS] Card list response received:\n%s\n\n", string(cardListResponseJson))

		if cardListResponse.Result.ResultStatus == "S" {
			cardCount := len(cardListResponse.CardList)
			log.Printf("[SUCCESS] Card inquiry successful - %d card(s) found\n", cardCount)

			if cardCount > 0 {
				log.Println("[INFO] Card details:")
				for i, card := range cardListResponse.CardList {
					log.Printf("  Card %d:\n", i+1)
					log.Printf("    Masked Card No: %s\n", card.MaskedCardNo)
					log.Printf("    Account Number: %s\n", card.AccountNumber)
				}
			} else {
				log.Println("[INFO] User has no cards bound to their account")
			}
		} else if cardListResponse.Result.ResultStatus == "F" {
			log.Printf("[ERROR] Card inquiry failed: %s\n", cardListResponse.Result.ResultMessage)
			log.Printf("[ERROR] Result code: %s\n", cardListResponse.Result.ResultCode)
		} else if cardListResponse.Result.ResultStatus == "U" {
			log.Printf("[WARNING] Card inquiry status unknown: %s\n", cardListResponse.Result.ResultMessage)
		}

		log.Println("[SUCCESS] Returning card list to frontend")
		log.Println("=================================================================\n")

		return ctx.JSON(cardListResponse)
	})
}

File: backend-node/src/api/inquiryEndpoint.js
const { applyToken, inquiryUserCardList } = require('../alipay');

function initInquiryEndpoint(router) {
    // Endpoint to exchange auth code for access token (specifically for card inquiry)
    router.post('/users/inquiry-cards/apply-token', async (req, res) => {
        try {
            const { auth_code } = req.body;

            if (!auth_code) {
                return res.status(400).json({ error: 'auth_code is required' });
            }

            console.log('=================================================================');
            console.log('STARTING TOKEN EXCHANGE FOR CARD INQUIRY');
            console.log('=================================================================');
            console.log(`[INFO] Auth code received from frontend`);

            const tokenResponse = await applyToken(auth_code);

            console.log('[SUCCESS] Token response received:');
            console.log(JSON.stringify(tokenResponse, null, 2));
            console.log('');

            if (tokenResponse.result.resultCode !== 'SUCCESS') {
                console.log(`[ERROR] Invalid token response: ${tokenResponse.result.resultMessage}`);
                return res.status(400).json({
                    error: 'Invalid token response: ' + tokenResponse.result.resultMessage
                });
            }

            console.log('[INFO] Token exchange successful');
            console.log(`[INFO] Customer ID: ${tokenResponse.customerId}`);
            console.log(`[INFO] Access token obtained (valid until: ${tokenResponse.accessTokenExpiryTime})`);
            console.log('[SUCCESS] Returning access token to frontend');
            console.log('=================================================================\n');

            return res.json({
                accessToken: tokenResponse.accessToken,
                customerId: tokenResponse.customerId,
                accessTokenExpiryTime: tokenResponse.accessTokenExpiryTime
            });

        } catch (err) {
            console.log(`[ERROR] Token exchange endpoint error: ${err.message}`);
            console.log('=================================================================\n');
            return res.status(500).json({ error: err.message });
        }
    });

    // Endpoint to get user card list using access token
    router.post('/users/inquiry-cards', async (req, res) => {
        try {
            const { accessToken } = req.body;

            if (!accessToken) {
                return res.status(400).json({ error: 'accessToken is required' });
            }

            console.log('=================================================================');
            console.log('STARTING USER CARD LIST INQUIRY');
            console.log('=================================================================');
            console.log('[INFO] Access token received from frontend');
            console.log('[INFO] Calling Alipay+ inquiryUserCardList API...');

            const cardListResponse = await inquiryUserCardList(accessToken);

            console.log('[SUCCESS] Card list response received:');
            console.log(JSON.stringify(cardListResponse, null, 2));
            console.log('');

            if (cardListResponse.result.resultStatus === 'S') {
                const cardCount = cardListResponse.cardList ? cardListResponse.cardList.length : 0;
                console.log(`[SUCCESS] Card inquiry successful - ${cardCount} card(s) found`);

                if (cardCount > 0) {
                    console.log('[INFO] Card details:');
                    cardListResponse.cardList.forEach((card, index) => {
                        console.log(`  Card ${index + 1}:`);
                        console.log(`    Masked Card No: ${card.maskedCardNo}`);
                        console.log(`    Account Number: ${card.accountNumber}`);
                    });
                } else {
                    console.log('[INFO] User has no cards bound to their account');
                }
            } else if (cardListResponse.result.resultStatus === 'F') {
                console.log(`[ERROR] Card inquiry failed: ${cardListResponse.result.resultMessage}`);
                console.log(`[ERROR] Result code: ${cardListResponse.result.resultCode}`);
            } else if (cardListResponse.result.resultStatus === 'U') {
                console.log(`[WARNING] Card inquiry status unknown: ${cardListResponse.result.resultMessage}`);
            }

            console.log('[SUCCESS] Returning card list to frontend');
            console.log('=================================================================\n');

            return res.json(cardListResponse);

        } catch (err) {
            console.log(`[ERROR] Card inquiry endpoint error: ${err.message}`);
            console.log('=================================================================\n');
            return res.status(500).json({ error: err.message });
        }
    });
}

module.exports = { initInquiryEndpoint };
